### PR TITLE
Remove positions from paths

### PR DESCRIPTION
**Pull request by lpw25**
Currrently, paths contain the integer position of a module field within the runtime representation of the module:

```ocaml
type t =
    Pident of Ident.t
  | Pdot of t * string * int
  | Papply of t * t
```

Since `Path.t`s appear within types, module types etc. they must remain valid through the various manipulations of the type system. Unfortunately, the positions do not remain valid under module substitution -- which leads to various soundness bugs:

```ocaml
        OCaml version 4.06.0

# module type S = sig
  module M : sig
    module A : sig end
    module B : sig end
  end
  module N = M.A
end

module Foo = struct
  module B = struct let x = 0 end
  module A = struct let x = "hello" end
end

module Bar : S with module M := Foo = struct module N = Foo.A end

let () = print_string Bar.N.x;;
                              Segmentation fault
```

```ocaml
        OCaml version 4.06.0

# module M : sig
  module N : sig
    module A : sig val x : string end
    module B : sig val x : int end
  end
  module F (X : sig module A = N.A end) : sig val s : string end
end = struct
  module N = struct
    module B = struct let x = 0 end
    module A = struct let x = "hello" end
  end
  module F (X : sig module A : sig val x : string end end) = struct
    let s = X.A.x
  end
end

module N = M.F(struct module A = M.N.A end)

let () = print_string N.s;;
                                    Segmentation fault
```

Rather than trying to change module substitution to take account of path indices, which appears quite difficult and will become even more difficult as the module system evolves, this PR removes the positions from paths. Instead we add an `address` type to the `Env` module:

```ocaml
type address =
  | Aident of Ident.t
  | Adot of address * int
```

along with functions for finding the "address" of a given path:

```ocaml
val find_value_address: Location.t -> Path.t -> t -> address
val find_module_address: Location.t -> Path.t -> t -> address
val find_class_address: Location.t -> Path.t -> t -> address
val find_constructor_address: Location.t -> Path.t -> t -> address
```

These functions are used in `bytecomp` to translate paths to lambda code. Since we previously had to call `normalize_path` at that point anyway, this is a pretty simple change to make.

### CODE REVIEW

**Comment by gasche** [comment]
If you read PRs through email and you were, like me, confused by the absence of an `int` argument in the `Path.t`  type given in Leo's message, this was a typo which is now fixed. Before the PR, `Path.t` is defined as

```ocaml
type t =
    Pident of Ident.t
  | Pdot of t * string * int
  | Papply of t * t
```

while the PR changes it to just (only `Pdot` changes, dropping the `int` argument):

```ocaml
type t =
    Pident of Ident.t
  | Pdot of t * string
  | Papply of t * t
```

**Comment by shindere** [comment]
Thanks, @gasche !

I do read PRs by e-mail and never think about checking the GitHub page
to see whether something has changed. I realise I should but still can't
remember doing so.

**Comment by lpw25** [comment]
Yeah, sorry about that. I should remember to make a comment when correcting particularly confusing typos.

**Comment by shindere** [comment]
Leo White (2018/02/14 10:23 +0000):
> Yeah, sorry about that. I should remember to make a comment when
> correcting particularly confusing typos.

No, no problem. I think the real issue is the interface. Perhaps it
would help to have a parameter to receive mails also when a PR or a
comment gets modified. With a diff of the modification. :)

**Comment by let-def** [comment]
`.depend` seems out of date.

**Comment by Drup** [comment]
Exposing the position in the path itself always seemed particularly dubious to me, so I'm happy to see that go away. 

Before I really dive into the code, can you give some indications on what this `module_presence` that is now sparkled everywhere is for ? Is it still only for module aliases ?

**Comment by lpw25** [comment]
To make the addresses a little easier to calculate I moved the presence information for module aliases from the module type (i.e. `Mty_alias(Mta_present, path)`) to the signature (i.e. `(Sig_module(id, Mta_present, Mty_alias(path))`). This also enabled me to fully remove absent aliases from the runtime representation -- previously they would get a slot containing `()` that was never used for anything.

**Comment by let-def** [comment]
From a first review, to me it looks like a clear improvement (I remember doing sometime wrong things with path positions when playing with modular implicits) and the patch *seems* correct but I don't understand all the parts.

**Comment by gasche** [comment]
Maybe you want to point at the parts you don't understand, and they could benefit from a comment :-)

File: typing/types.ml
Comment (Drup) [inline typing/types.ml:171]: I'm not very familiar with compilation of inlined record. Why is the path now needed here?

File: bytecomp/translcore.ml
Comment (Drup) [inline bytecomp/translcore.ml:134]: On various occasion where a typechecking function synthesized a new desc, you now add it to the environment before returning it, where previously it was just returned. I suppose it is to make sure it has an address, but I'm not really sure why it is ok to add new things to the env that were not added previously.

Also, in that specific case, it's added only to the expression part, not the pattern one.

File: typing/env.ml
Comment (Drup) [inline typing/env.ml]: When normalizing, you consider modules that are aliases but have an address. I failed to find where you introduce aliases that have an address different than `None`. Where does that happen?

File: toplevel/toploop.ml
Comment (Drup) [inline toplevel/toploop.ml]: Maybe factorize that code a little bit? :)

**Review by Drup — COMMENTED** [review]
The main commit is quite difficult to review (mostly trivial or loosely related changes spread out over 50+ files, and the occasional subtle thing). I would have preferred to have the the purely trivial parts split out from the subtle ones, and then squash before merge. That would have made it much easier to review.

But apart from that, I think this is a clear improvement. The handling of address is much more obvious than before, and this removes some weird invariants from the typedtree. The code is also a tiny bit cleaner.

I mostly looked at the typechecking-related parts, as I'm not an expert in `bytecomp/`, but it all looks pretty good to me.

File: typing/types.ml
Comment (trefis) [reply typing/types.ml:171]: My guess would be: to avoid doing things [like this](https://github.com/ocaml/ocaml/pull/1610/files?diff=unified#diff-e8899086c6b3526d6bc4453b4ad5395cL1304).
That is: unless I missunderstood, it is a cleanup unrelated to the rest of this PR.

File: bytecomp/translcore.ml
Comment (trefis) [reply bytecomp/translcore.ml:134]: > On various occasion where a typechecking function synthesized a new desc, you now add it to the environment before returning it, where previously it was just returned. I suppose it is to make sure it has an address, but I'm not really sure why it is ok to add new things to the env that were not added previously.

Was "typechecking" a typo here?
Also, regardless of why it is now added in the environment (though I think I agree with the reason you propose), I think it's more correct (in this particular instance, i.e. in `push_defaults`) to add it to the environment than not.

> Also, in that specific case, it's added only to the expression part, not the pattern one.

This is similar to what happens in the typechecker, if you look at `Typecore.type_cases` you'll see that the pattern variables (i.e. the bindings produced by the pattern) only get added to the environment of the body, not of the pattern.

File: typing/types.ml
Comment (Drup) [reply typing/types.ml:171]: Ah, I missed that piece of code. That's a very valid reason, although if it is, it really should be in a different commit.

File: typing/env.ml
Comment (lpw25) [reply typing/env.ml]: So if a module is present (i.e. it's presence is `Mta_present` in a module type) then it will be given an address (see `store_module` and `components_of_module_maker`). This can happen for aliases than are produced via strengthening (see `strengthen_decl` which can produce aliases, and `strengthen_sig` which then attaches these aliases to a presence that can be `Mta_present`).

File: toplevel/toploop.ml
Comment (lpw25) [reply toplevel/toploop.ml]: Done

**Comment by lpw25** [comment]
Sorry for the slow response. I've finally found time to look at this again. I've rebased the code and addressed @Drup's comments.

**Comment by Drup** [comment]
My small concerns have been addressed. I think we need a review from someone familiar with bytecomp now.

**Comment by damiendoligez** [comment]
Since this fixes a soundness bug, I would like it to see it move forward.
@trefis would you do a formal review?

**Comment by lpw25** [comment]
Could we hold off for a week or so? I've got an improved version implemented as part of my transparent ascription patch, and I'm planning to extract it into this PR.

**Comment by lpw25** [comment]
Ok I've pushed the improved version and rebased it up to trunk. Should now be ready for review.

**Comment by trefis** [comment]
I'm having a look.

File: typing/env.ml
Comment (trefis) [inline typing/env.ml:2118]: To be clear:
```
scope (create_scoped ~scope:0 "foo") == 0
scope (create_local "foo")           == highest_scope (== 100_000_000)
```
So that's an actual change, somewhat unrelated to this PR.
But I believe it is fine: we only check the scope of type constructors, we should never check the scope of value paths.

File: typing/types.mli
Comment (trefis) [inline typing/types.mli]: Should we rename these to `Mp_present | Mp_absent`?

File: bytecomp/translmod.ml
Comment (trefis) [inline bytecomp/translmod.ml]: I think this deserves a comment.
Apparently `p1` in these cases would be `-1`, because it's not a runtime component (cf. `includemod.ml`), so we can't really go have a look in `v2`.

File: bytecomp/translmod.ml
Comment (trefis) [inline bytecomp/translmod.ml]: You could drop the `cont` parameter of `store_alias`, have it return the `Lprim` and simplify this last line to: `make_sequence store_alias aliases`.

File: bytecomp/matching.ml
Comment (trefis) [inline bytecomp/matching.ml]: Is there a reason to use `Location.none` over `loc` here?

File: typing/env.ml
Comment (trefis) [inline typing/env.ml]: That feels wrong: it "works", but it's weird.
Here is the type of the function after this change:
```ocaml
val diff_keys: ('a -> bool) -> 'b t -> ('a * 'c) t -> Ident.t list
```
Before that it was:
```ocaml
val diff_keys: ('a -> bool) -> 'b t -> 'a t -> Ident.t list
```
which seems more natural.

I think it's the caller of this function that should be changed, while this function should remain untouched. (And the function appears to be called only once, so it shouldn't be too tedious to update)

File: typing/env.ml
Comment (trefis) [inline typing/env.ml:411]: Like for `TycompTbl.diff_keys` the changes to this function are weird, and inconsistent with the other changes in this module: this function is making the assumption that the `'a` in `current : Ident.tbl (* in [t] *)` and in `components: 'a list NameMap.t (* in [opened] *)` is a tuple:
```ocaml
val find_all: string -> ('a * 'b) t -> (Path.t * 'a) list
```
But there's no reason to. And, for example, I believe you wouldn't be able to call `find_all` on the `modtypes` field of the environment with this implementation.
It's the callers that should change.

**Review by trefis — COMMENTED** [review]
This overall looks good (and correct) modulo the few comments I left here and there.

**Comment by lpw25** [comment]
Ok, I've addressed the review comments and fixed the failing tests.

**Comment by trefis** [comment]
I forgot to say it during my review: but you updated toploop,topdirs,etc. But not the opt versions. And you'll need to :)

File: bytecomp/translclass.ml
Comment (let-def) [inline bytecomp/translclass.ml]: Very minor: strange indentation for `::inh_init` (the cons applies to the triplet, not to `envs`), they could all be put on the same line.

File: bytecomp/translclass.ml
Comment (let-def) [inline bytecomp/translclass.ml:462]: Another minor remark about something that I found a bit confusing at first:
- the result of `transl` functions has the form `(path, path_lam, obj_init)` where the last two values are lambda expressions (the first to access the path at runtime, the second one is the function that initializes an object)
- the accumulator of inherited initialisations is a list where elements have the form `(obj_init, path, path_lam)`, with the same meaning as above.
Reading closely I don't think that any mistake is caused by this permutation.

Comment (lpw25) [reply bytecomp/translclass.ml:462]: Fixed

File: bytecomp/translclass.ml
Comment (lpw25) [reply bytecomp/translclass.ml]: Fixed

**Comment by lpw25** [comment]
I've updated ocamlnat and addressed @let-def's comments.

**Review by let-def — APPROVED** [review]

**Comment by lpw25** [comment]
Rebased. Merging once CI passes.

File: debugger/loadprinter.ml
Comment (alainfrisch) [inline debugger/loadprinter.ml:97]: While rebasing #2131 on top of this one, I found this spot.  Does the comment above about PR#7258 still applies?  Does the new code handle module aliases properly?   (There is a similar stuff in toploop.ml and opttoploop.ml.)

Comment (lpw25) [reply debugger/loadprinter.ml:97]: I should have removed that comment. The `find_foo_address` functions already handle module aliases so there is no need to normalize paths before calling them.

Comment (damiendoligez) [reply debugger/loadprinter.ml:97]: In a case like that, you can push a direct commit to remove the comment, no need for a new PR.

Comment (lpw25) [reply debugger/loadprinter.ml:97]: Done

**Pull request merged**
